### PR TITLE
[tcp] add TCPlp's FreeBSD-derived protocol code, with explanatory comments

### PR DESCRIPTION
**Pull request by samkumar**
This is the eighth in a series of pull requests (after #6744, #6770, #6790, #6857, #6885, #6925, and #6926) bringing TCPlp to OpenThread. I'm making a series of a small pull requests so that it's easier for the OpenThread community to review the code. If you'd like to preview a working version of TCPlp in OpenThread, see #6650.

This pull request adds TCPlp's TCP protocol logic, which is derived from the TCP protocol logic in the FreeBSD Operating System. The FreeBSD code had to be refactored to work for TCPlp, and I have added extensive comments explaining the code that was deleted and changed. The changes are primarily modifications to work with TCPlp's data buffering (see #6926), removal of features that depend on dynamic memory allocation to save memory (SYN cache, compressed TIME-WAIT state) and removal of features that aren't relevant in this setting (e.g., TCP Segmentation Offloading, IPv4 support).

### CODE REVIEW

**Review by abtink — APPROVED** [review]
Thanks @samkumar. 
Overall looks good (though I must admit I did not delve into the code logic in detail). 

Few style questions/suggestion below.

File: third_party/tcplp/bsdtcp/tcp.h
Comment (samkumar) [reply third_party/tcplp/bsdtcp/tcp.h:108]: The different alignments are present in the original FreeBSD code. I believe that it does indicate a dpendency of some sort. For example, after the option number (e.g.,TCPOPT_EOL), the option length (e.g., TCPOLEN_EOL) is indented. I will update this to reflect the alignment used in the latest FreeBSD code (https://github.com/freebsd/freebsd-src/blob/main/sys/netinet/tcp.h) --- now that I look at it, it seems that the first few options in this block acquired an extra indent, somehow, which I will fix.

File: third_party/tcplp/bsdtcp/tcp.h
Comment (samkumar) [reply third_party/tcplp/bsdtcp/tcp.h:169]: The alignment in this block appears to have been introduced by me somehow (the original FreeBSD code doesn't have it), so I will fix it. Thanks for noticing this.

File: third_party/tcplp/bsdtcp/tcp_const.h
Comment (samkumar) [reply third_party/tcplp/bsdtcp/tcp_const.h:65]: Thanks for noticing this; I will fix it.

File: third_party/tcplp/bsdtcp/sys/queue.h
Comment (samkumar) [reply third_party/tcplp/bsdtcp/sys/queue.h:304]: Hm, I'm not sure what happened in this file --- it looks like the original FreeBSD file had all of the `\`s aligned. I'll look into it and see how to fix it.

File: third_party/tcplp/bsdtcp/tcp.h
Comment (samkumar) [reply third_party/tcplp/bsdtcp/tcp.h:169]: The alignment in this block appears to have been introduced by me somehow (the original FreeBSD code doesn't have it), so I will fix it. Thanks for noticing this.

File: third_party/tcplp/bsdtcp/tcp_const.h
Comment (samkumar) [reply third_party/tcplp/bsdtcp/tcp_const.h:65]: Thanks for noticing this; I will fix it.

File: third_party/tcplp/bsdtcp/sys/queue.h
Comment (samkumar) [reply third_party/tcplp/bsdtcp/sys/queue.h:304]: I looked into this, and it turns out that if you set the tab size to 8 spaces, then all the `\`s are aligned. The reason they appear unaligned is that GitHub is displaying tabs with a width of 4 spaces.

Any thoughts as to what the right solution is for this?

Comment (samkumar) [reply third_party/tcplp/bsdtcp/sys/queue.h:304]: If you look at the file directly in GitHub, it renders correctly, with tabs set to 8 spaces: https://github.com/samkumar/openthread/blob/feature-tcp-tcplp-freebsd/third_party/tcplp/bsdtcp/sys/queue.h

So I think that this is merely an artifact of GitHub's pull request interface, and that the `\`s will be properly aligned once this pull request is merged. So I think that nothing needs to be done about this at the moment.

Thoughts?

File: third_party/tcplp/bsdtcp/tcp.h
Comment (samkumar) [reply third_party/tcplp/bsdtcp/tcp.h:169]: It turns out that the FreeBSD code actually has this questionable alignment, because it uses tabs for some lines and spaces for others? I fixed it in this pull request though.

File: third_party/tcplp/bsdtcp/tcp.h
Comment (samkumar) [reply third_party/tcplp/bsdtcp/tcp.h:108]: It turns out that the FreeBSD code actually has this questionable alignment, because it uses tabs for some lines and spaces for others? I fixed it in this pull request though.

**Comment by samkumar** [comment]
@abtink Thanks for the comments. I've pushed a commit addressing them. So far, I've left sys/queue.h the same because when viewed with a tab width of 8 spaces all of the `\` characters are aligned. If you view the code directly, outside of a diff, it displays correctly. I'm open to changing the file if you think we should, so let me know if you'd like to see any specific changes there.

**Comment by abtink** [comment]
Thanks @samkumar. Since it is third party code, I think it's perfectly fine to keep the style and format as they are in the original code. I would say the much more important aspect when dealing with third party code is to ensure ease of applying future patches (git merges, cherry-picks, etc).  So keeping the code as close to the original may be actually better.  😃 
For any new file (new code) it makes sense to follow a common style.

**Comment by samkumar** [comment]
Thanks @abtink for your guidance. `tcp_const.h` is a new file that I added, so I kept my changes there. For the remaining files, I discarded my changes to keep it identical to the FreeBSD versions, for ease of applying patches, as you suggested.

**Review by abtink — APPROVED** [review]
Thanks. 👍

**Comment by jwhui** [comment]
@samkumar , to address the compiler warnings, can we resolve them by stripping compiler flags?

Example:
https://github.com/openthread/openthread/blob/b4c76b3e170a729f9bdb57ddaa6d05b73e541ce2/third_party/mbedtls/CMakeLists.txt#L45

**Comment by samkumar** [comment]
@jwhui I made some preliminary efforts in that direction, but I found that the specific flags that caused warnings somewhat differ from compiler to compiler, and some compilers give the warnings even after stripping compiler flags. I'll take another look at this evening and post here with any specific errors that I can't eliminate this way. Perhaps the right thing to do is to to eliminate as many warnings as we can via the compiler flags, and then manually fix the ones that can't be eliminated that way.

**Comment by samkumar** [comment]
@jwhui I managed to find flags to add and filter across that work across the tested compiler versions to suppress the warnings. At this point, I believe that all tests pass. If you and @simonlingoogle are OK with the current version, then I can squash the commits to prepare for merging this pull request.

**Review by simonlingoogle — APPROVED** [review]
LGTM :+1: 

Thanks

**Review by jwhui — APPROVED** [review]

**Comment by jwhui** [comment]
> @jwhui I managed to find flags to add and filter that work across the tested compiler versions to suppress the warnings. At this point, I believe that all tests pass. If you and @simonlingoogle are OK with the current version, then I can squash the commits to prepare for merging this pull request.

@samkumar , I can go ahead and squash-merge this if you like. Let me know if you want to make any other changes before merging.

**Comment by samkumar** [comment]
@jwhui I'm happy with merging this as is. I just squashed and rebased, so it should all be good to go.

File: third_party/tcplp/bsdtcp/tcp_subr.c
Comment (canisLupus1313) [inline third_party/tcplp/bsdtcp/tcp_subr.c:232]: @samkumar this part of code is showing problems on static analysis of code. tt_ipgen has size of 40 bytes but otMessageInfo is size of 44 bytes. was this intentional to zero out tcphdr tt_t.

**Pull request merged**
